File: sim/golden_models/models/AfeAfe2256Model.cpp
#include "golden_models/models/AfeAfe2256Model.h"

#include "golden_models/core/TestVectorIO.h"

namespace fpd::sim {

namespace {

constexpr uint32_t kAfe2256MinTLine = 5120U;

std::vector<uint16_t> MakeAfe2256Row(uint32_t profile, uint32_t seed) {
    std::vector<uint16_t> row(256U);
    const uint16_t gain = static_cast<uint16_t>((profile + 1U) * 7U);
    for (uint32_t index = 0; index < row.size(); ++index) {
        row[index] = static_cast<uint16_t>((seed + index * gain) & 0xFFFFU);
    }
    return row;
}

}  // namespace

void AfeAfe2256Model::reset() {
    afe_start_ = 0;
    config_req_ = 0;
    cfg_tline_ = kAfe2256MinTLine;
    cfg_cic_en_ = 0;
    cfg_cic_profile_ = 0;
    cfg_pipeline_en_ = 0;
    cfg_tp_sel_ = 0;
    cfg_nchip_ = 1;
    afe_ready_ = 0;
    config_done_ = 0;
    dout_window_valid_ = 0;
    fclk_expected_ = 0;
    line_count_ = 0;
    tline_error_ = 0;
    pipeline_latency_rows_ = 0;
    previous_row_.clear();
    current_row_.clear();
    cycle_count_ = 0;
}

void AfeAfe2256Model::step() {
    config_done_ = 0;
    tline_error_ = (cfg_tline_ < kAfe2256MinTLine) ? 1U : 0U;
    if (config_req_ != 0U && afe_ready_ == 0U) {
        afe_ready_ = 1U;
        config_done_ = 1U;
    }

    if (afe_start_ != 0U && afe_ready_ != 0U && tline_error_ == 0U) {
        dout_window_valid_ = 1U;
        fclk_expected_ = 1U;
        ++line_count_;
        current_row_ = MakeAfe2256Row(cfg_cic_profile_, cfg_tp_sel_ + (cfg_nchip_ << 4U));
        if (cfg_pipeline_en_ != 0U) {
            ++pipeline_latency_rows_;
            if (!previous_row_.empty()) {
                current_row_ = previous_row_;
            }
            previous_row_ = MakeAfe2256Row(cfg_cic_profile_, cfg_tp_sel_ + (cfg_nchip_ << 4U) + 1U);
        }
        if (cfg_cic_en_ != 0U) {
            for (auto& sample : current_row_) {
                sample = static_cast<uint16_t>((sample * (cfg_cic_profile_ + 1U)) & 0xFFFFU);
            }
        }
        if (line_count_ >= cfg_tline_) {
            line_count_ = 0U;
            dout_window_valid_ = 0U;
            fclk_expected_ = 0U;
        }
    } else {
        dout_window_valid_ = 0U;
        fclk_expected_ = 0U;
        line_count_ = 0U;
    }
    ++cycle_count_;
}

void AfeAfe2256Model::set_inputs(const SignalMap& inputs) {
    afe_start_ = GetScalar(inputs, "afe_start", afe_start_);
    config_req_ = GetScalar(inputs, "config_req", config_req_);
    cfg_tline_ = GetScalar(inputs, "cfg_tline", cfg_tline_);
    cfg_cic_en_ = GetScalar(inputs, "cfg_cic_en", cfg_cic_en_);
    cfg_cic_profile_ = GetScalar(inputs, "cfg_cic_profile", cfg_cic_profile_);
    cfg_pipeline_en_ = GetScalar(inputs, "cfg_pipeline_en", cfg_pipeline_en_);
    cfg_tp_sel_ = GetScalar(inputs, "cfg_tp_sel", cfg_tp_sel_);
    cfg_nchip_ = GetScalar(inputs, "cfg_nchip", cfg_nchip_);
}

SignalMap AfeAfe2256Model::get_outputs() const {
    return {
        {"config_done", config_done_},
        {"afe_ready", afe_ready_},
        {"dout_window_valid", dout_window_valid_},
        {"fclk_expected", fclk_expected_},
        {"tline_error", tline_error_},
        {"pipeline_latency_rows", pipeline_latency_rows_},
        {"line_pixels", current_row_},
    };
}

std::vector<Mismatch> AfeAfe2256Model::compare(const SignalMap& rtl_outputs) const {
    return CompareSignalMaps(cycle(), get_outputs(), rtl_outputs);
}

void AfeAfe2256Model::generate_vectors(const std::string& output_dir) {
    TestVectorFile vectors;
    vectors.module_name = "afe_afe2256";
    vectors.spec_name = "SPEC-FPD-006";
    vectors.clock_name = "mclk";
    vectors.signal_inputs = {
        "afe_start", "config_req", "cfg_tline", "cfg_cic_en",
        "cfg_cic_profile", "cfg_pipeline_en", "cfg_tp_sel"
    };
    vectors.signal_outputs = {
        "config_done", "afe_ready", "dout_window_valid", "fclk_expected",
        "tline_error", "pipeline_latency_rows"
    };

    reset();
    set_inputs({{"config_req", 1U}, {"cfg_tline", 5120U}});
    step();
    vectors.vectors.push_back({cycle(), {{"config_req", 1U}, {"cfg_tline", 5120U}}, get_outputs()});

    set_inputs({
        {"afe_start", 1U},
        {"cfg_tline", 5120U},
        {"cfg_cic_en", 1U},
        {"cfg_cic_profile", 2U},
        {"cfg_pipeline_en", 1U},
        {"cfg_tp_sel", 1U},
    });
    step();
    vectors.vectors.push_back({cycle(), {
        {"afe_start", 1U}, {"cfg_tline", 5120U}, {"cfg_cic_en", 1U},
        {"cfg_cic_profile", 2U}, {"cfg_pipeline_en", 1U}, {"cfg_tp_sel", 1U}
    }, get_outputs()});

    WriteHexVectors(vectors, output_dir + "/afe_afe2256_vectors.hex");
    WriteBinaryVectors(vectors, output_dir + "/afe_afe2256_vectors.bin");
}

}  // namespace fpd::sim

File: sim/viewer/src/FpdSimViewer/Models/AfeAfe2256Model.cs
using FpdSimViewer.Models.Core;

namespace FpdSimViewer.Models;

public sealed class AfeAfe2256Model : GoldenModelBase
{
    private const uint kAfe2256MinTLine = 5120U;

    private uint _afeStart;
    private uint _configReq;
    private uint _cfgTLine = kAfe2256MinTLine;
    private uint _cfgCicEn;
    private uint _cfgCicProfile;
    private uint _cfgPipelineEn;
    private uint _cfgTpSel;
    private uint _cfgNChip = 1U;
    private uint _afeReady;
    private uint _configDone;
    private uint _doutWindowValid;
    private uint _fclkExpected;
    private uint _lineCount;
    private uint _tlineError;
    private uint _pipelineLatencyRows;
    private ushort[] _previousRow = [];
    private ushort[] _currentRow = [];

    public override void Reset()
    {
        _afeStart = 0U;
        _configReq = 0U;
        _cfgTLine = kAfe2256MinTLine;
        _cfgCicEn = 0U;
        _cfgCicProfile = 0U;
        _cfgPipelineEn = 0U;
        _cfgTpSel = 0U;
        _cfgNChip = 1U;
        _afeReady = 0U;
        _configDone = 0U;
        _doutWindowValid = 0U;
        _fclkExpected = 0U;
        _lineCount = 0U;
        _tlineError = 0U;
        _pipelineLatencyRows = 0U;
        _previousRow = [];
        _currentRow = [];
        CycleCount = 0;
    }

    public override void Step()
    {
        _configDone = 0U;
        _tlineError = _cfgTLine < kAfe2256MinTLine ? 1U : 0U;

        if (_configReq != 0U && _afeReady == 0U)
        {
            _afeReady = 1U;
            _configDone = 1U;
        }

        if (_afeStart != 0U && _afeReady != 0U && _tlineError == 0U)
        {
            _doutWindowValid = 1U;
            _fclkExpected = 1U;
            _lineCount++;
            _currentRow = MakeRow(_cfgCicProfile, _cfgTpSel + (_cfgNChip << 4));
            if (_cfgPipelineEn != 0U)
            {
                _pipelineLatencyRows++;
                if (_previousRow.Length != 0)
                {
                    _currentRow = _previousRow;
                }

                _previousRow = MakeRow(_cfgCicProfile, _cfgTpSel + (_cfgNChip << 4) + 1U);
            }

            if (_cfgCicEn != 0U)
            {
                for (var index = 0; index < _currentRow.Length; index++)
                {
                    _currentRow[index] = (ushort)((_currentRow[index] * (_cfgCicProfile + 1U)) & 0xFFFFU);
                }
            }

            if (_lineCount >= _cfgTLine)
            {
                _lineCount = 0U;
                _doutWindowValid = 0U;
                _fclkExpected = 0U;
            }
        }
        else
        {
            _doutWindowValid = 0U;
            _fclkExpected = 0U;
            _lineCount = 0U;
        }

        CycleCount++;
    }

    public override void SetInputs(SignalMap inputs)
    {
        _afeStart = SignalHelpers.GetScalar(inputs, "afe_start", _afeStart);
        _configReq = SignalHelpers.GetScalar(inputs, "config_req", _configReq);
        _cfgTLine = SignalHelpers.GetScalar(inputs, "cfg_tline", _cfgTLine);
        _cfgCicEn = SignalHelpers.GetScalar(inputs, "cfg_cic_en", _cfgCicEn);
        _cfgCicProfile = SignalHelpers.GetScalar(inputs, "cfg_cic_profile", _cfgCicProfile);
        _cfgPipelineEn = SignalHelpers.GetScalar(inputs, "cfg_pipeline_en", _cfgPipelineEn);
        _cfgTpSel = SignalHelpers.GetScalar(inputs, "cfg_tp_sel", _cfgTpSel);
        _cfgNChip = SignalHelpers.GetScalar(inputs, "cfg_nchip", _cfgNChip);
    }

    public override SignalMap GetOutputs()
    {
        return new SignalMap
        {
            ["afe_ready"] = _afeReady,
            ["config_done"] = _configDone,
            ["dout_window_valid"] = _doutWindowValid,
            ["fclk_expected"] = _fclkExpected,
            ["line_count"] = _lineCount,
            ["tline_error"] = _tlineError,
            ["pipeline_latency_rows"] = _pipelineLatencyRows,
            ["previous_row"] = _previousRow,
            ["current_row"] = _currentRow,
            ["line_pixels"] = _currentRow,
        };
    }

    private static ushort[] MakeRow(uint profile, uint seed)
    {
        var row = new ushort[256];
        var gain = (ushort)((profile + 1U) * 7U);
        for (var index = 0; index < row.Length; index++)
        {
            row[index] = (ushort)((seed + (index * gain)) & 0xFFFFU);
        }

        return row;
    }
}
